File: skills/project-init/assets/run_pipeline.py
#!/usr/bin/env python3
"""Run the project Python scripts in order and save an automatic run record."""

from __future__ import annotations

from datetime import datetime
from hashlib import sha256
from pathlib import Path
import json
import os
import platform
import subprocess
import sys


ROOT = Path(__file__).resolve().parent
if not (ROOT / "02_code").is_dir():
    raise SystemExit("run_pipeline.py must be located and run at the project root")

# List preparation and formal analysis separately. The readiness gate runs before analysis.
preparation_scripts = [
    ROOT / "02_code" / "01_data_cleaning.py",
]
analysis_scripts: list[Path] = []
scripts = [*preparation_scripts, *analysis_scripts]
readiness_helper = ROOT / "02_code" / "vendored" / "data_readiness.py"
missing_scripts = [path.relative_to(ROOT).as_posix() for path in scripts if not path.is_file()]
if analysis_scripts and not readiness_helper.is_file():
    missing_scripts.append(readiness_helper.relative_to(ROOT).as_posix())
if missing_scripts:
    raise SystemExit(f"formal analysis scripts are missing: {', '.join(missing_scripts)}")

started = datetime.now().astimezone()
run_id = f"{started.strftime('%Y%m%dT%H%M%S%z')}_{os.getpid()}"
runs = ROOT / "results" / "runs"
runs.mkdir(parents=True, exist_ok=True)
log_path = runs / f"{run_id}.log"
environment_path = runs / f"{run_id}-environment.txt"
environment = dict(os.environ)
environment["EPI_RUN_ID"] = run_id
exit_code = 0


def run_step(
    script: Path,
    log,
    *,
    arguments: list[str] | None = None,
    label: str | None = None,
) -> int:
    relative = label or script.relative_to(ROOT).as_posix()
    log.write(f"\n===== {relative} =====\n")
    completed = subprocess.run(
        [sys.executable, str(script), *(arguments or [])],
        cwd=ROOT,
        env=environment,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        text=True,
        encoding="utf-8",
        errors="replace",
    )
    log.write(completed.stdout)
    return completed.returncode


with log_path.open("w", encoding="utf-8") as log:
    for script in preparation_scripts:
        exit_code = run_step(script, log)
        if exit_code:
            break
    if not exit_code and analysis_scripts:
        exit_code = run_step(
            readiness_helper,
            log,
            arguments=["--check"],
            label="data readiness gate",
        )
    if not exit_code:
        for script in analysis_scripts:
            exit_code = run_step(script, log)
            if exit_code:
                break

environment_lines = [
    f"python={sys.version.replace(chr(10), ' ')}",
    f"executable={sys.executable}",
    f"platform={platform.platform()}",
]
for lock_name in ("requirements.txt", "pyproject.toml", "uv.lock", "poetry.lock"):
    lock_path = ROOT / lock_name
    if lock_path.is_file():
        environment_lines.append(f"declared_dependency_file={lock_name}")
environment_path.write_text("\n".join(environment_lines) + "\n", encoding="utf-8")

output_roots = [ROOT / name for name in ("results", "03_tables", "04_figures", "paper", "05_reports")]
files: dict[str, str] = {}
for output_root in output_roots:
    if not output_root.is_dir():
        continue
    for path in output_root.rglob("*"):
        if path.is_file() and runs not in path.parents:
            files[path.relative_to(ROOT).as_posix()] = sha256(path.read_bytes()).hexdigest()

run_record = {
    "run_id": run_id,
    "started_at": started.isoformat(timespec="seconds"),
    "finished_at": datetime.now().astimezone().isoformat(timespec="seconds"),
    "status": "success" if exit_code == 0 else "failed",
    "exit_code": exit_code,
    "command": f"{Path(sys.executable).name} run_pipeline.py",
    "scripts": [path.relative_to(ROOT).as_posix() for path in scripts],
    "log": log_path.relative_to(ROOT).as_posix(),
    "environment": environment_path.relative_to(ROOT).as_posix(),
    "hash_algorithm": "sha256",
    "files": files,
}
record_text = json.dumps(run_record, ensure_ascii=False, indent=2) + "\n"
(runs / f"{run_id}.json").write_text(record_text, encoding="utf-8")
latest_tmp = runs / ".latest.json.tmp"
latest_tmp.write_text(record_text, encoding="utf-8")
os.replace(latest_tmp, runs / "latest.json")

if exit_code:
    raise SystemExit(f"project scripts failed; inspect {log_path.relative_to(ROOT)}")
print(f"project scripts completed; run ID: {run_id}")

File: skills/project-init/assets/run_pipeline.R
# Run the project R scripts in order and save an automatic run record.
if (!dir.exists("02_code")) stop("请从项目根运行 run_pipeline.R")

# 数据准备和正式统计分析分别列出；第一项分析前会检查数据就绪状态。
preparation_scripts <- c(
  "02_code/01_data_cleaning.R"
)
analysis_scripts <- character()
scripts <- c(preparation_scripts, analysis_scripts)
readiness_helper <- "02_code/vendored/data_readiness.R"
missing_scripts <- scripts[!file.exists(scripts)]
if (length(analysis_scripts) && !file.exists(readiness_helper)) {
  missing_scripts <- c(missing_scripts, readiness_helper)
}
if (length(missing_scripts)) {
  stop("正式分析脚本缺失：", paste(missing_scripts, collapse = "，"))
}

json_escape <- function(value) {
  value <- gsub("\\\\", "\\\\\\\\", as.character(value))
  value <- gsub('"', '\\\\"', value, fixed = TRUE)
  value <- gsub("\r", "\\\\r", value, fixed = TRUE)
  gsub("\n", "\\\\n", value, fixed = TRUE)
}
json_string <- function(value) paste0('"', json_escape(value), '"')
json_array <- function(values) paste0("[", paste(vapply(values, json_string, character(1)), collapse = ", "), "]")

started <- Sys.time()
run_id <- paste0(format(started, "%Y%m%dT%H%M%S%z"), "_", Sys.getpid())
dir.create("results/runs", recursive = TRUE, showWarnings = FALSE)
log_path <- file.path("results/runs", paste0(run_id, ".log"))
environment_path <- file.path("results/runs", paste0(run_id, "-environment.txt"))
status <- 0L

run_script <- function(script, run_id, args = character()) {
  previous_run_id <- Sys.getenv("EPI_RUN_ID", unset = NA_character_)
  Sys.setenv(EPI_RUN_ID = run_id)
  on.exit(
    if (is.na(previous_run_id)) Sys.unsetenv("EPI_RUN_ID") else Sys.setenv(EPI_RUN_ID = previous_run_id),
    add = TRUE
  )
  system2(
    file.path(R.home("bin"), "Rscript"),
    c("--vanilla", script, args),
    stdout = TRUE,
    stderr = TRUE
  )
}

run_logged_step <- function(script, run_id, args = character(), label = script) {
  output <- run_script(script, run_id, args)
  script_status <- attr(output, "status")
  if (is.null(script_status)) script_status <- 0L
  cat("\n===== ", label, " =====\n", paste(output, collapse = "\n"), "\n",
      file = log_path, append = TRUE, sep = "")
  as.integer(script_status)
}

for (script in preparation_scripts) {
  step_status <- run_logged_step(script, run_id)
  if (step_status != 0L) {
    status <- step_status
    break
  }
}

if (status == 0L && length(analysis_scripts)) {
  step_status <- run_logged_step(
    readiness_helper,
    run_id,
    args = "--check",
    label = "data readiness gate"
  )
  if (step_status != 0L) status <- step_status
}

if (status == 0L) {
  for (script in analysis_scripts) {
    step_status <- run_logged_step(script, run_id)
    if (step_status != 0L) {
      status <- step_status
      break
    }
  }
}

writeLines(capture.output(sessionInfo()), environment_path, useBytes = TRUE)
output_roots <- c("results", "03_tables", "04_figures", "paper", "05_reports")
files <- unlist(lapply(output_roots[dir.exists(output_roots)], function(root) {
  list.files(root, recursive = TRUE, full.names = TRUE, all.files = FALSE)
}), use.names = FALSE)
files <- files[file.exists(files) & !dir.exists(files)]
files <- files[!grepl("^results/runs/", gsub("\\\\", "/", files))]
hashes <- if (length(files)) tools::md5sum(files) else character()
file_lines <- if (length(files)) vapply(seq_along(files), function(index) {
  sprintf("    %s: %s", json_string(gsub("\\\\", "/", files[index])), json_string(unname(hashes[index])))
}, character(1)) else character()
if (length(file_lines) > 1L) file_lines[-length(file_lines)] <- paste0(file_lines[-length(file_lines)], ",")

run_record <- c(
  "{",
  sprintf("  \"run_id\": %s,", json_string(run_id)),
  sprintf("  \"started_at\": %s,", json_string(format(started, "%Y-%m-%dT%H:%M:%S%z"))),
  sprintf("  \"finished_at\": %s,", json_string(format(Sys.time(), "%Y-%m-%dT%H:%M:%S%z"))),
  sprintf("  \"status\": %s,", json_string(if (status == 0L) "success" else "failed")),
  sprintf("  \"exit_code\": %d,", status),
  "  \"command\": \"Rscript --vanilla run_pipeline.R\",",
  sprintf("  \"scripts\": %s,", json_array(gsub("\\\\", "/", scripts))),
  sprintf("  \"log\": %s,", json_string(gsub("\\\\", "/", log_path))),
  sprintf("  \"environment\": %s,", json_string(gsub("\\\\", "/", environment_path))),
  "  \"hash_algorithm\": \"md5\",",
  "  \"files\": {",
  file_lines,
  "  }",
  "}"
)
record_path <- file.path("results/runs", paste0(run_id, ".json"))
writeLines(run_record, record_path, useBytes = TRUE)
latest_tmp <- file.path("results/runs", ".latest.json.tmp")
writeLines(run_record, latest_tmp, useBytes = TRUE)
if (file.exists("results/runs/latest.json")) unlink("results/runs/latest.json")
if (!file.rename(latest_tmp, "results/runs/latest.json")) stop("无法更新最近一次运行记录")

if (status != 0L) stop("项目脚本运行失败；请检查 ", log_path, call. = FALSE)
message("项目脚本运行完成；运行编号：", run_id)
